kernel/solo5.h

```diff
@@ -23,135 +23,248 @@
+    /*
+     * The operation cannot be completed at this time. Retrying an identical
+     * operation at a later time may succeed.
+     */
+    SOLO5_R_AGAIN,
+    /*
+     * Invalid argument.
+     */
+    SOLO5_R_EINVAL,
+    /*
+     * The operation failed due to an unspecified error.
+     */
+    SOLO5_R_EUNSPEC
+} solo5_result_t;
 
 /*
- * Boot-time information passed to application at entry.
+ * Application entry point.
+ *
+ * The application is passed a pointer (*info) to a statically allocated
+ * (struct solo5_start_info) with start of day information; any data in this
+ * structure should be treated as read-only.
+ *
+ * The application receives the (info->heap_start, info->heap_size) of a
+ * single, contiguous, non-executable, read/write memory region it may use for
+ * heap or stack(s) as it sees fit.  At entry, the application is provided with
+ * an initial stack growing down from (info->heap_start + info->heap_size).
+ *
+ * The application MUST NOT make any further assumptions about memory layout,
+ * including where executable code or static data are located in memory.
+ *
+ * Returning from this function is equivalent to calling solo5_exit(<return
+ * value>).
  */
-struct solo5_boot_info {
+struct solo5_start_info {
     const char *cmdline;
     uintptr_t heap_start;
     size_t heap_size;
 };
 
+int solo5_app_main(const struct solo5_start_info *info);
+
 /*
- * Application entry point.
+ * Exit the application, returning (status) to the host if possible.
  *
- * Returning from this function is equivalent to calling solo5_exit(ret).
+ * Status values of 255 and above are reserved for use by Solo5.
  */
-int solo5_app_main(const struct solo5_boot_info *bi);
+#define SOLO5_EXIT_SUCCESS	0
+#define SOLO5_EXIT_FAILURE	1
+#define SOLO5_EXIT_ABORT	255
+
+void solo5_exit(int status) __attribute__((noreturn));
 
 /*
- * Network I/O.
+ * Time.
  */
 
 /*
- * Send a single network packet (*data) of size (n). Returns -1 on failure, 0
- * on success.
+ * Solo5 type for time values, with nanosecond precision.
  */
-int solo5_net_write_sync(uint8_t *data, int n);
+typedef uint64_t solo5_time_t;
 
 /*
- * Receive a single network packet into (*data) up to a maximum size of (*n).
- * Returns 0 on success and size of received packet (*n), -1 on failure or if
- * no packets are available.
+ * Returns monotonic time since an unspecified period in the past.
  */
-int solo5_net_read_sync(uint8_t *data, int *n);
+solo5_time_t solo5_clock_monotonic(void);
 
 /*
- * Returns a pointer to the network MAC address, formatted as a C string
- * XX:XX:XX:XX:XX:XX.
+ * Returns wall clock time in UTC since the UNIX epoch (1970-01-01).
  */
-char *solo5_net_mac_str(void);
+solo5_time_t solo5_clock_wall(void);
 
 /*
- * Block I/O.
+ * Suspends execution of the application until either:
+ *
+ *   a) monotonic time reaches (deadline), or
+ *   b) solo5_net_recv() would succeed.
+ *
+ * Returns true if solo5_net_recv() will succeed, otherwise false.
```

Comment:
in 142 and 144, you should replace `solo5_net_recv` with `solo5_net_read`

Reply:
@hannesm Done, thanks for reviewing!